src/proxysql.rs

```diff
@@ -0,0 +1,249 @@
+use chrono::{DateTime, Local};
+use mysql::{prelude::Queryable, Conn, OptsBuilder};
+
+use crate::{
+    config,
+    hosts::{Host, HostStatus},
+    messages,
+    queries::Query,
+};
+
+const MIRROR_QUERY_TOKEN: &str = "Mirror by readyset scheduler at";
+const DESTINATION_QUERY_TOKEN: &str = "Added by readyset scheduler at";
+pub struct ProxySQL {
+    readyset_hostgroup: u16,
+    warmup_time_s: u16,
+    conn: mysql::Conn,
+    hosts: Vec<Host>,
+}
+
+impl ProxySQL {
+    /// This function is used to create a new ProxySQL struct.
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - A reference to a config::Config containing the configuration for the ProxySQL connection.
+    ///
+    /// # Returns
+    ///
+    /// A new ProxySQL struct.
+    pub fn new(config: &config::Config) -> Self {
+        let mut conn = Conn::new(
+            OptsBuilder::new()
+                .ip_or_hostname(Some(config.proxysql_host.as_str()))
+                .tcp_port(config.proxysql_port)
+                .user(Some(config.proxysql_user.as_str()))
+                .pass(Some(config.proxysql_password.as_str()))
+                .prefer_socket(false),
+        )
+        .expect("Failed to create ProxySQL connection");
+
+        let query = format!(
+            "SELECT hostname, port, status, comment FROM mysql_servers WHERE hostgroup_id = {}",
+            config.readyset_hostgroup
+        );
+        let results: Vec<(String, u16, String, String)> = conn.query(query).unwrap();
+        let hosts = results
+            .into_iter()
+            .filter_map(|(hostname, port, status, comment)| {
+                if comment.to_lowercase().contains("readyset") {
+                    Some(Host::new(hostname, port, status, config))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<Host>>();
+
+        ProxySQL {
+            conn,
+            readyset_hostgroup: config.readyset_hostgroup,
+            warmup_time_s: config.warmup_time_s.unwrap_or(0),
+            hosts,
+        }
+    }
+
+    /// This function is used to add a query rule to ProxySQL.
+    ///
+    /// # Arguments
+    ///
+    /// * `query` - A reference to a Query containing the query to be added as a rule.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating if the rule was added successfully.
+    pub fn add_as_query_rule(&mut self, query: &Query) -> Result<bool, mysql::Error> {
+        let datetime_now: DateTime<Local> = Local::now();
+        let date_formatted = datetime_now.format("%Y-%m-%d %H:%M:%S");
```

Comment:
Do we also need to write the time zone here?

Reply:
No need to complicate here. This will always be run on the same ProxySQL server, so no timezone diff.

Reply:
I'm only bringing this up because I see we're decoding some time zone below with %z.  That seems okay?